spec/services/rate_sleeper_spec.rb

```diff
@@ -0,0 +1,59 @@
+require 'rails_helper'
+
+RSpec.describe RateSleeper do
+  describe ".run" do
+    context "with jobs per second = 8" do
+      let(:rate_sleeper) {
+        RateSleeper.new(jobs_per_second: 8)
+      }
+
+      context "with almost 0 execution time" do
+        it "sleeps for about 0.125 seconds" do
+          #measuring the duration of yield takes a tiny amount
+          #of time so it won't be quite 0.125
+
+          allow(rate_sleeper).to receive(:sleep)
+          expect(rate_sleeper).to receive(:sleep).with(be_between(0.124, 0.126))
+
+          rate_sleeper.run {}
+        end
+      end
+
+      context "with 0.05 seconds execution time" do
+        it "sleeps for about 0.07 seconds" do
+          expect(rate_sleeper).to receive(:sleep).with(be_between(0.06, 0.08))
+
+          rate_sleeper.run {
+            Kernel.sleep(0.05)
+          }
+        end
+      end
+
+      context "with execution time slower than the allowed rate" do
+        it "doesn't sleep" do
+          expect(rate_sleeper).not_to receive(:sleep)
+
+          rate_sleeper.run {
+            Kernel.sleep(0.13)
+          }
+        end
+      end
+    end
+
+    context "with jobs per second 20" do
```

Comment:
Just wondering what this context "tests" given we are already exercising the code's maths (and hurting my head manually calculating it all) in https://github.com/alphagov/email-alert-api/pull/233/files#diff-7adf437dfb818ed2ba40a94bbea07125R5 ? Is it just that we can pass in different `jobs_per_second` and that will affect the resulting sleep times?

Reply:
yep, it's just testing that the configuration parameter actually configures the thing correctly